File: repo/frontend/src/features/books/isbn.ts
/**
 * ISBN rules, the same as the API's: spaces and hyphens are dropped, a final
 * "x" becomes "X", and the result must be a valid ISBN-10 or ISBN-13.
 */

export function normalizeIsbn(value: string): string {
  return value.replace(/[\s-]/g, "").toUpperCase()
}

/** True when a normalized value has the shape of an ISBN-10 or ISBN-13. */
export function hasIsbnShape(isbn: string): boolean {
  return /^\d{9}[\dX]$/.test(isbn) || /^\d{13}$/.test(isbn)
}

/** True when a normalized ISBN-10 or ISBN-13 has a correct check digit. */
export function hasValidIsbnChecksum(isbn: string): boolean {
  const digits = [...isbn].map((char) => (char === "X" ? 10 : Number(char)))
  if (/^\d{9}[\dX]$/.test(isbn)) {
    // Weights 10 down to 1; the sum is a multiple of 11.
    return digits.reduce((sum, digit, index) => sum + digit * (10 - index), 0) % 11 === 0
  }
  if (/^\d{13}$/.test(isbn)) {
    // Weights alternate 1 and 3; the sum is a multiple of 10.
    return digits.reduce((sum, digit, index) => sum + digit * (index % 2 ? 3 : 1), 0) % 10 === 0
  }
  return false
}

File: repo/backend/app/services/isbn.py
"""ISBN normalization and check-digit rules for ISBN-10 and ISBN-13."""

import re

_SEPARATORS = re.compile(r"[\s-]+")
_ISBN10 = re.compile(r"[0-9]{9}[0-9X]")
_ISBN13 = re.compile(r"[0-9]{13}")
# What a search term looks like once separators are removed, when it could be part of an ISBN.
_ISBN_FRAGMENT = re.compile(r"[0-9]+X?")


def normalize_isbn(raw: str) -> str:
    """Removes spaces and hyphens and upper-cases a final x: "0-441-17271-7" -> "0441172717"."""
    return _SEPARATORS.sub("", raw).upper()


def has_isbn_shape(isbn: str) -> bool:
    """True for 10 characters (digits, a final X allowed) or 13 digits."""
    return bool(_ISBN10.fullmatch(isbn) or _ISBN13.fullmatch(isbn))


def check_digit_matches(isbn: str) -> bool:
    """True when a normalized ISBN-10 or ISBN-13 passes its checksum."""
    if _ISBN10.fullmatch(isbn):
        # Weights 10 down to 1; X stands for 10.
        values = [10 if char == "X" else int(char) for char in isbn]
        return sum((10 - i) * value for i, value in enumerate(values)) % 11 == 0
    if _ISBN13.fullmatch(isbn):
        return sum(int(char) * (3 if i % 2 else 1) for i, char in enumerate(isbn)) % 10 == 0
    return False


def isbn_search_fragment(term: str) -> str | None:
    """The normalized digits of a search term that could be part of an ISBN, otherwise None."""
    fragment = normalize_isbn(term)
    return fragment if _ISBN_FRAGMENT.fullmatch(fragment) else None
